### PR TITLE
Add separate dialog to notebook charts for specifying chart options

### PR DESCRIPTION
**Pull request by corivera**
Breaking out the chart options into a separate dialog helps save some space in the notebook outputs. Chart option controls are still rendered inline in the query editor.

Capture

### CODE REVIEW

**Comment by corivera** [comment]
Doesn't look like anything gets focused by default in the dialog. Also there's an interesting bug where hitting the Enter key when focused on the Close button also reopens the dialog again, maybe since the open chart button is highlighted at the same time.

**Comment by chlafreniere** [comment]
> Doesn't look like anything gets focused by default in the dialog. Also there's an interesting bug where hitting the Enter key when focused on the Close button also reopens the dialog again, maybe since the open chart button is highlighted at the same time.

Oh wow ok. Let's try to fix those issues as much as we can. I know that you've got limited time, but anything that you can do accessibility-wise to save us in the future would be much appreciated.

File: src/sql/workbench/contrib/charts/browser/chartView.ts [hunk 1]
```diff
@@ -99,8 +102,13 @@ export class ChartView extends Disposable implements IPanelView {
 		this._createInsightAction = this._instantiationService.createInstance(CreateInsightAction);
 		this._copyAction = this._instantiationService.createInstance(CopyAction);
 		this._saveAction = this._instantiationService.createInstance(SaveImageAction);
+		this._configureChartAction = this._instantiationService.createInstance(ConfigureChartAction, this);
```
Comment (chlafreniere) [inline src/sql/workbench/contrib/charts/browser/chartView.ts]: might want to only create this if this._renderOptionsInline is false

**Comment by chlafreniere** [comment]
Hey @alanrenmsft if you have any cycles, would be helpful if you could take a look and try to spot any accessibility issues.

File: src/sql/workbench/contrib/charts/browser/actions.ts [hunk 2]
```diff
@@ -106,6 +108,28 @@ export class CreateInsightAction extends Action {
 	}
 }
 
+export class ConfigureChartAction extends Action {
+	public static ID = 'chartview.configureChart';
+	public static LABEL = localize('configureChartLabel', "Configure Chart");
+	public static ICON = 'filterLabel';
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/charts/browser/actions.ts]: >filterLabel [](start = 23, length = 11)

Why are you using the filter icon here?

File: src/sql/workbench/contrib/charts/browser/configureChartDialog.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+	public render() {
+		super.render();
+		attachModalDialogStyler(this, this._themeService);
+
+		let closeButton = this.addFooterButton(localize('optionsDialog.close', "Close"), () => this.cancel());
+
+		attachButtonStyler(closeButton, this._themeService);
+		this._register(this._themeService.onThemeChange(e => this.updateTheme(e)));
+		this.updateTheme(this._themeService.getTheme());
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/charts/browser/configureChartDialog.ts]: If you're not going to change anything might as well not bother attaching the listeners

File: src/sql/workbench/contrib/charts/browser/configureChartDialog.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+	public open() {
+		this.show();
+	}
+
+	public render() {
+		super.render();
+		attachModalDialogStyler(this, this._themeService);
+
+		let closeButton = this.addFooterButton(localize('optionsDialog.close', "Close"), () => this.cancel());
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/charts/browser/configureChartDialog.ts]: `configureChartDialog.close`

File: src/sql/workbench/contrib/charts/test/browser/chartView.test.ts [hunk 5]
```diff
@@ -31,5 +31,5 @@ function createChartView(): ChartView {
 	const instantiationService = new TestInstantiationService();
 	const notificationService = new TestNotificationService();
 	instantiationService.stub(IThemeService, themeService);
-	return new ChartView(contextViewService, themeService, instantiationService, notificationService);
+	return new ChartView(true, contextViewService, themeService, instantiationService, notificationService);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/charts/test/browser/chartView.test.ts]: Suggest making these tests test with both values of renderInline

File: src/sql/workbench/services/objectExplorer/browser/media/objectTypes/objecttypes.css [hunk 6]
```diff
@@ -9,6 +9,12 @@
 	background: url("Table.svg") center center no-repeat;
 }
 
+.vs .codicon.table,
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/media/objectTypes/objecttypes.css]: >codicon [](start = 5, length = 7)

Hm, why is this change necessary? I'd think we'd be using the same classes for displaying the action icons. And Kevin fixed a bunch of issues a while back where we had to update the icons to use codicon instead so I feel like this is just a case of these being missed...

File: src/sql/workbench/contrib/charts/browser/chartView.ts [hunk 7]
```diff
@@ -68,7 +69,7 @@ export class ChartView extends Disposable implements IPanelView {
 	/** parent container */
 	private container: HTMLElement;
 	/** container for the options controls */
-	private optionsControl: HTMLElement;
+	public readonly optionsControl: HTMLElement;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/charts/browser/chartView.ts]: It seems like with your changes it would be good to pull the options out into its own class so we can have more direct control over it since it's not always going to be directly tied to the chart UI-layout-wise

File: src/sql/workbench/contrib/charts/browser/configureChartDialog.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+		attachButtonStyler(closeButton, this._themeService);
+		this._register(this._themeService.onThemeChange(e => this.updateTheme(e)));
+		this.updateTheme(this._themeService.getTheme());
+	}
+
+	protected renderBody(container: HTMLElement) {
+		container.appendChild(this._chart.optionsControl);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/charts/browser/configureChartDialog.ts]: Here seems like a good place to pick the element to focus on. This is where pulling the options element into its own class will help since you can add a getter there to get an appropriate element (likely just the first one in the top section of options)

File: src/sql/workbench/services/objectExplorer/browser/media/objectTypes/objecttypes.css (hunk 6 above)
Comment (corivera) [reply src/sql/workbench/services/objectExplorer/browser/media/objectTypes/objecttypes.css]: Some things use .icon, and other use .codicon. Our dialogs use .codicon, but the table icon was only defined for .icon before.

Comment (Charles-Gagnon) [reply src/sql/workbench/services/objectExplorer/browser/media/objectTypes/objecttypes.css]: Yes, but we really should align to use codicon everywhere. We had icon because that's what vs code originally had but they changed everything over to codicon a while back and to be consistent I think it'd be easiest to just also make sure we're using codicon everywhere.

There's an argument to be made for making an adsicon or something to differentiate our icons vs theirs - but honestly I think that's just adding extra confusion for no real gain.

Anyways that's all outside the scope of this PR - my original point was why is this necessary for this PR? I'm guessing it's because the action is adding the codicon class to the class list but not the icon class?

File: src/sql/workbench/contrib/charts/browser/actions.ts (hunk 2 above)
Comment (corivera) [reply src/sql/workbench/contrib/charts/browser/actions.ts]: The only gear icon I could find was the class **codicon-settings-gear**, which overrides some font and size styling.

File: src/sql/workbench/services/objectExplorer/browser/media/objectTypes/objecttypes.css (hunk 6 above)
Comment (corivera) [reply src/sql/workbench/services/objectExplorer/browser/media/objectTypes/objecttypes.css]: Items in the object explorer use .icon, for example.

File: src/sql/workbench/contrib/charts/browser/chartView.ts (hunk 7 above)
Comment (corivera) [reply src/sql/workbench/contrib/charts/browser/chartView.ts]: Currently the dialog is just an alternate display method. Rendering just the optionsControl element elsewhere seemed to be a less disruptive way of achieving that.

File: src/sql/workbench/contrib/charts/browser/actions.ts (hunk 2 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/charts/browser/actions.ts]: Couldn't you override those stylings yourself in the css for this?

If that's not possible for some reason we should make our own then. The icon as it currently stands is definitely not going to be something people expect/understand to mean settings.

File: src/sql/workbench/services/objectExplorer/browser/media/objectTypes/objecttypes.css (hunk 6 above)
Comment (corivera) [reply src/sql/workbench/services/objectExplorer/browser/media/objectTypes/objecttypes.css]: This is a styling change to align the chart toolbar with the design mockups. The show table button needed a table icon, but we use .codicon for buttons in dialogs.

File: src/sql/workbench/contrib/charts/browser/actions.ts (hunk 2 above)
Comment (corivera) [reply src/sql/workbench/contrib/charts/browser/actions.ts]: Looks like the icons are drawn from a custom font, so overriding the font is something we can't do. They probably aren't meant to be used for labels. I'll see if I can extract the settings icon somehow.

https://github.com/microsoft/azuredatastudio/blob/master/src/vs/base/browser/ui/codiconLabel/codicon/codicon.css#L8

Comment (corivera) [reply src/sql/workbench/contrib/charts/browser/actions.ts]: Added a new icon.

**Comment by corivera** [comment]
@chlafreniere For the issue where the dialog immediately opens again when hitting enter on the Close button, it looks like we handle dialog close on KEY_DOWN, and running actionbar actions with KEY_UP. So after closing the dialog, we run the openDialog action when lifting off the enter key. I'll have to address that issue separately, since it would affect all actionbars.

**Comment by corivera** [comment]
Added a tracking item for the actionbar KEY_UP issue here: https://github.com/microsoft/azuredatastudio/issues/9515

**Comment by chlafreniere** [comment]
> @chlafreniere For the issue where the dialog immediately opens again when hitting enter on the Close button, it looks like we handle dialog close on KEY_DOWN, and running actionbar actions with KEY_UP. So after closing the dialog, we run the openDialog action when lifting off the enter key. I'll have to address that issue separately, since it would affect all actionbars.
> 
> Fortunately, the Escape key also exits the dialog.

Thanks, @corivera. Makes sense to me.

**Comment by corivera** [comment]
@Charles-Gagnon I implemented your accessibility recommendations. Could you take another look here?

File: src/sql/workbench/contrib/charts/browser/configureChartDialog.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+				(inputControlItems[0] as HTMLElement).focus();
+			}
+		}
+	}
+
+	public render() {
+		super.render();
+		attachModalDialogStyler(this, this._themeService);
+
+		let closeButton = this.addFooterButton(localize('optionsDialog.close', "Close"), () => this.close());
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/charts/browser/configureChartDialog.ts]: This should probably be configureChartDialog.close
